core/templates/dev/head/expressions/expressionSyntaxTree.js

```diff
@@ -0,0 +1,420 @@
+        }
+        if (args.length >= expectedNum && args.length <= expectedMax) {
+          return;
+        }
+        throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
+      };
+
+      var _verifyArgsHaveType = function(args, type) {
+        for (var i = 0; i < args.length; i++) {
+          if (args[i] != type) {
+            throw new ExprWrongArgTypeError(
+              originalValue, typeof originalValue, 'Number');
+          }
+        }
+        return true;
+      };
+
+      var evaluateExpression = function(expression, envs, evaluate) {
+        return evaluateParseTree(expressionParserService.parse(expression),
+          envs, evaluate);
+      };
+
+      var evaluateParseTree = function(parsed, envs, evaluate) {
+        return evaluate(parsed, envs.concat(system));
+      };
+
+      /**
+       * Looks up a variable of the given name in the env. Here the variable
+       * can be system or user defined functions and parameters, as well as
+       * system operators.
+       * @param {string} name The name to look up.
+       * @param {!Array.<!Object>} envs Represents a nested name space
+       *     environment to look up the name in. The first element is looked up
+       *     first (i.e. has higher precedence).
+       * @throws {ExprUndefinedVarError} The named variable was not found in
+       *     the given environment.
+       */
+      var lookupEnvs = function(name, envs) {
+        // Parameter value look up.
+        var value;
+        if (envs.some(function(env) {
+          if (env.hasOwnProperty(name)) {
+            value = env[name];
+            return true;
+          }
+          return false;
+        })) {
+          return value;
+        }
+
+        throw new ExprUndefinedVarError(name, envs);
+      };
+
+      // Coerces the argument to a Number, and throws an error if the result
+      // is NaN.
+      var _coerceToNumber = function(originalValue) {
+        var coercedValue = (+originalValue);
+        if (!isNaN(coercedValue)) {
+          return coercedValue;
+        }
+        throw new ExprWrongArgTypeError(
+          originalValue, typeof originalValue, 'Number');
+      };
+
+      // Coerces all values in the given argument array to Number, and throws
+      // an error if the result is NaN.
+      var _coerceAllArgsToNumber = function(args) {
+        for (var i = 0; i < args.length; i++) {
+          args[i] = _coerceToNumber(args[i]);
+        }
+        return args;
+      };
+
+      // NOTE TO DEVELOPERS: When adding a new reserved word to this object,
+      //   please first ensure that existing explorations do not use this
+      //   parameter name. Also, to prevent future explorations using it,
+      //   modify feconf.INVALID_PARAMETER_NAMES accordingly.
+      // TODO(kashida): Document all operators input and output contracts.
+
+      var system = {
+    '#': {
+      type: function(args, envs) {
+        return lookupEnvs(args + '', envs);
```

Comment:
RIght now it is being sent a map that maps the name of the var to the type, which is why the lookup looks like this, so that is what this code is doing.